rfcs/0000-ret-cont-recursive-nix.md

```diff
@@ -0,0 +1,245 @@
+```nix
+{ stdenv, pkgs, nix }:
+
+stdenv.mkDerivation {
+  name = "foo";
+  version = "1.2.3";
+
+  src = ...;
+
+  nativeBuildInputs = [ nix ];
+  NIX_PATH = "nixpkgs=${pkgs.path}";
+
+  __recursive = true;
+
+  outputs = [ "drv" ];
+
+  doConfigure = false;
+  doBuild = false;
+
+  installPhase = ''
+    mv $(nix-instantiate -E '((import <nixpkgs> {}).callPackage ./. {}).'"$o") $drv
+  '';
+}
+```
+Note how in this case we don't need to do any "post-processing" of the produced derivation.
+When the outer derivation can just "become" the inner derivation, explicitly copying the derivation outputs like before becomes unnecessary.
+
+So why prefer this variation of the standard design?
+I've always been concerned with the ease of which someone can just "nix-build ...; nix-build ...; nix-build ..." within a derivation with recursive Nix.
+This creates a linear chain of dependencies, which isn't terribly performant: shorter critical paths are crucial for parallelism and incrementality and this fails with both.
+Building derivations is less convenient, but makes linear chains and the proper dependency graph *equally* less convenient, removing the perverse incentive.
+And in general, dynamism in the dependency graph, which is the essence of what recursive Nix provides, is only a feature of last resort, so making it more difficult across the board isn't concerning.
+
+Additionally, see https://github.com/edolstra/nix/commit/1a27aa7d64ffe6fc36cfca4d82bdf51c4d8cf717 for Eelco's draft implementation of recursive Nix, and the Darwin sandboxing restrictions that make it a Linux-only feature.
+Sandboxing and Darwin are crucial to Nix today, and we shouldn't sacrifice either of them.
+With "ret-cont" recursive Nix, actual builds are never nested, so we don't need any fancy constraints on the derivation "runtime" (i.e. the code that actually performs and isolates builds).
+Furthermore, we can skip needing to talk to the daemon by just producing a local store:
+```nix
+stdenv.mkDerivation {
+  # ...
+  outputs = [ "drv" "store" ];
+  installPhase = ''
+    mv $(nix-instantiate --store $store -E '((import <nixpkgs> {}).callPackage ./. {}).'"$o") $drv
+  '';
+}
+```
+This further simplifies the implementation.
+Derivations remain built exactly as today, with only logic *between* building steps that is entirely platform-agnostic changing.
+
+# Detailed design
+[design]: #detailed-design
+
+Derivations today build outputs, and are associated to those outputs.
+We extend the derivation language by allowing a derivation to indicate their output is more derivations, and ultimately be associated with one of *those* derivations's associated outputs.
+Derivations that that do so indicate this with some special attribute, say `__recursive`.
+Such derivations must have two outputs, `store` and `drv`.
+`store` would be a local Nix store limited to just drvs and fixed output builds.
+`drv` would contain a symlink to one of the derivations in the store, the root.
```

Comment:
Note that ideally for recursive drvs you also have the actual outputs of the final derivation specified, so the nix-lang usage of them can just reference outputs appropriately

Reply:
Ah yes that's true.